Adds initial draft for headers topic

rwdougla opened the pull request:
resolves cplusplus/SG20#3

rwdougla committed:
WIP: Straw-man "Headers" for small-group session

Issue cplusplus/SG20#3

uvic-frodo commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,105 @@
+
+#### Background/Required Knowledge
+
+* All of the above.
```
Is "All of the above" on line 77 referring to all of the foundational background?  I think that it would be clearer to say this explicitly.  Incidentally, is it implied that all background for a topic at the foundational level is also required for the main level?  If so, then perhaps such a statement is unnecessary?

rwdougla committed:
WIP: progress from small group session

Issue cplusplus/SG20#3

rwdougla committed:
Progress from March US Small Group session

Would like to get cross-references in advance section to other modules which are coupled to headers, such as classes, templates, modules, etc.

vulder committed:
Merge branch 'master' into Issue3

vulder commented:
/ok-to-test

rwdougla committed:
Update headers.md

feeling of small-group was to omit advanced section, as each advanced topic seemed more relevant to a different module

vulder committed:
Merge branch 'master' into Issue3

diehlpk commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+* Usage of standard library code requires inclusion of various header files
+* Give examples of various header files and when they may be needed for inclusion
+* Include links to reference materials for finding out what headers are used for various feature sets
+* When using standard library facilities, can either use `using std::vector; vector<int> v;`, or `std::vector<int> v;`, or `using namespace std; vector<int> v;`, though modern practice strongly discourages invoking `using namespace std;`
```
That is a nice example but that does not relates to headers. I think it relates to templates and I am not sure if we should mention it there.

Thoughts?

vulder replied:
From what I see, this relates directly to headers. The reasoning this point tries to get across is the fact that `using namespace x`, in a header introduces all names within `x` into all files that include the specific header (even transitively). This not only makes it likely to have name clashes but also increases the chance that, for example, some usages of functions with the same name as is present in `x`, select a different (better fitting) overload, which can introduce subtle errors.

I assume, @rwdougla wanted to have this point covered when talking about headers, so that students are aware and do not misuse `using namespace x`. 

Maybe we should highlight the reasoning here so it's clearer.

diehlpk replied:
I agree but maybe elaborating would help.

vulder committed:
Merge branch 'master' into Issue3

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+## Module name: Headers
```
```suggestion
## C++ compilation model: Headers {#headers}
```

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+multiple distinct source files to share a single source of common 
+declarations. 
+
+### Foundational: Including code through standard library headers
```
```suggestion
### Foundational: Including code through standard library headers {#headers-found}
```

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+* Include links to reference materials for finding out what headers are used for various feature sets
+* When using standard library facilities, can either use `using std::vector; vector<int> v;`, or `std::vector<int> v;`, or `using namespace std; vector<int> v;`, though modern practice strongly discourages invoking `using namespace std;`
+
+### Main: Organizing function and class declarations for reuse
```
```suggestion
### Main: Organizing function and class declarations for reuse  {#headers-main}
```

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+
+A student should be able to:
+
+1. Include common standard library headers to get access to standard library features
```
```suggestion
1. include common standard library headers to get access to standard library features
```

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+
+A student should be able to:
+
+1. Create a declaration for an existing function, placed in a separate file
+2. Utilize double-quote inclusion to include a header from the same directory
+3. Utilize angular-bracket inclusion to include a header from standard library
+4. Explain the meaning of the one-definition rule and how it applies to headers
+5. Protect a header with include guards and explain why they are necessary
```
```suggestion
1. create a declaration for an existing function, placed in a separate file
2. utilize double-quote inclusion to include a header from the same directory
3. utilize angular-bracket inclusion to include a header from standard library
4. explain the meaning of the one-definition rule and how it applies to headers
5. protect a header with include guards and explain why they are necessary
```
alignment with other topic spellings

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,106 @@
+#### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
```
```suggestion
implementation-defined, unspecified, or undefined behavior._

* The include directly actually copies the header file into the file it is included in.

```

vulder reviewed (changes requested):
Nice, work.

@rwdougla What's the status of this topic, should we table it for a SG20 discussion? It seems quite ready.

diehlpk commented:
@rwdougla could you please have a look and accept the changes?

vulder commented:
@diehlpk I talked to @rwdougla. I'll take care of the PR on the weekend and clean in up :)

diehlpk commented:
@vulder Perfect. Thanks. We will work next week on the lambda pull request.

vulder committed:
Apply suggestions from code review

vulder committed:
Merge branch 'master' into Issue3

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,108 @@
+* Usage of standard library code requires inclusion of various header files
+* Give examples of various header files and when they may be needed for inclusion
+* Include links to reference materials for finding out what headers are used for various feature sets
+* When using standard library facilities, can either use `using std::vector; vector<int> v;`, or `std::vector<int> v;`, or `using namespace std; vector<int> v;`, though modern practice strongly discourages invoking `using namespace std;`
```
```suggestion
* When using standard library facilities, one can either use `using std::vector; vector<int> v;`, or `std::vector<int> v;`, or `using namespace std; vector<int> v;`, though modern practice strongly discourages invoking `using namespace std;`
```

vulder committed:
Apply suggestions from code review

vulder committed:
Merge branch 'master' into Issue3

vulder commented:
@diehlpk Please have a look, I cleaned up the few open points. If there are no further problems to address, I would mark the PR tentative ready for the next online SG20 meeting.

diehlpk reviewed (approved):
LGTM!

vulder committed:
Update sources/modules/compilation-model/headers.md

vulder commented on sources/modules/compilation-model/headers.md:
```diff
@@ -0,0 +1,107 @@
+* Circular dependencies can cause confusion. Care should be taken in identifying such
+
+#### Points to cover
+
```
After looking at the points to cover in the SG20 meeting, we found that the points to cover are small/less that what is part of the ASBATS. SG20 suggests to extend the points to cover to include content wise what is asked for in the ASBATS

vulder replied:
@rwdougla Hi, could you please have a quick look at this, so we can vote on this tentative PR in the next meeting.
